runtime/minor_gc.c

```diff
@@ -466,13 +472,28 @@ CAMLexport void caml_gc_dispatch (void)
   if (trigger != caml_young_alloc_start || caml_requested_major_slice){
     /* The minor heap is half-full, do a major GC slice. */
     caml_requested_major_slice = 0;
-    caml_young_trigger = caml_young_alloc_start;
-    caml_young_limit = caml_young_trigger;
+    set_trigger(caml_young_alloc_start);
     caml_major_collection_slice (-1);
     CAML_INSTR_TIME (tmr, "dispatch/major");
   }
 }
 
+/* Called by [Alloc_small] when [caml_young_ptr] reaches [caml_young_limit].
+   We have to either call memprof or the gc. */
+void caml_alloc_small_dispatch (tag_t tag, intnat wosize, int track)
+{
+  if (caml_young_ptr < caml_young_trigger){
+    caml_young_ptr += Whsize_wosize (wosize);
+    CAML_INSTR_INT ("force_minor/alloc_small@", 1);
+    caml_gc_dispatch ();
+    caml_young_ptr -= Whsize_wosize (wosize);
+  }
+  if(caml_young_ptr < caml_memprof_young_trigger){
+    if(track) caml_memprof_track_young((tag), (wosize));
+    else caml_memprof_renew_minor_sample();
```

Comment:
I think that I understand what this code does now. I wondered if it would be possible to always renew the sample here, and call `caml_memprof_track_young` after that -- if I understand correctly, you wouldn't need to renew at all in `caml_memprof_track_young` in that case. The reason why you are not doing this is that the beginning of `caml_memprof_track_young` needs access to the pre-renewal value of `caml_memprof_young_trigger`. Would it be nice to save this value here (or even the difference with the young pointer) and pass it as a parameter to `caml_memprof_track_young` instead?

Reply:
`caml_memprof_young_trigger` is not the only variable to save. We would also have to save `next_sample_frac_part`. I can do what you propose (I do something similar for native sampling), but this essentially requires to split `caml_memprof_track_young` in two...

Reply:
I see. I think the current version is reasonable, so it's probably not worth changing, although the split version could be conceptually nicer -- it splits the responsability between the part that maintains the evolving state, and the part that consumes it.